tianshou/data/batch.py

```diff
@@ -500,6 +503,35 @@ def __getitem__(self, index: str | IndexType) -> Any:
             return new_batch
         raise IndexError("Cannot access item from empty Batch object.")
 
+    def __eq__(self, other: Self) -> bool:  # type: ignore
+        this_dict = self.__dict__
+        other_dict = other.__dict__
+
+        if len(this_dict) != len(other_dict):
```

Comment:
This is a bit complicated. Could you pls extend the implementation of `to_dict` with `recursive=False`, and thereby allow converting a `Batch` into a pure `python-numpy-torch` object. Then you can probably use [DeepDiff](https://github.com/seperman/deepdiff) to implement equality check

Reply:
`DeepDiff` is great, but it seems that it doesn't work for `torch.Tensor`s. 

One solution would be to convert the torch tensors to numpy arrays before applying `DeepDiff`.

I am investigating now whether there is some way to overcome this problem.

Reply:
I tried to pass a custom `hasher` to `DeepDiff` that would convert a `torch.Tensor` to `np.ndarray` but w/o success.

Then, I skimmed through the code and saw that `Batch` already has a convenient method `to_numpy()` which changes all `torch.Tensor` to `np.ndarray` albeit in-place.

So, what we can do here is call `to_numpy()` before converting to `dict` and comparing differences with `DeepDiff`.

I think we need to implement a non in-place version of `to_numpy()` here because:

- when testing for inequality the tensors of the two objects may be converted to numpy and the user is unaware
- it is weird that the object is changing internally when testing for equality.

I also saw that there is a non in-place `to_numpy()` function in `tianshou.data.utils.converter`. Using this one though requires some refactoring because currently it imports `Batch` and this leads to a circular import when using it in `Batch`.

Reply:
> Could you pls extend the implementation of to_dict with recursive=False

I might have misinterpreted this comment.

So, the previous implementation of `to_dict` was implicitly recursive. The extended version now has `recurse=False` by default. But I think by default it should remain recursive as this is what you would expect if you'd like to convert some object to dict.

Regarding nomenclature, I agree that `recursive` is better than `recurse`. I think I went with `recurse` to have some consistency like in `Batch.is_empty`.

Reply:
P.S. IMO, the name of the method `Batch.to_numpy()` is not reflecting what it does. A better name would be `convert_tensor_values_to_numpy`?